File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/8. 1012.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
let T = +inputArr[0]
let inputIdx = 1

// 출력 변수
let resultStr = ''

// N, M, K. 쿼리마다 갱신함
let N, M, K

// 농장 배열. 쿼리마다 갱신함
let farm = []

// 방문여부 배열. 쿼리마다 갱신함
let isVisit = []

while (T--) {
  const [m, n, k] = inputArr[inputIdx++].trim().split(' ').map(Number)

  N = n
  M = m
  K = k

  farm = Array.from({length: N}, () => Array.from({length: M}).fill(0))

  for (let i = 0; i < K; i++) {
    const [X, Y] = inputArr[inputIdx++].trim().split(' ').map(Number)
    farm[Y][X] = 1
  }

  isVisit = Array.from({length: N}, () => Array.from({length: M}).fill(false))

  let result = 0
  for (let row = 0; row < N; row++) {
    for (let col = 0; col < M; col++) {
      if (farm[row][col] === 1 && !isVisit[row][col]) {
        result++
        dfs(row, col)
      }
    }
  }
  resultStr += result + '\n'
}

// 출력
process.stdout.write(resultStr)

function dfs(row, col) {
  isVisit[row][col] = true

  if (row > 0 && farm[row - 1][col] === 1 && !isVisit[row - 1][col]) {
    dfs(row - 1, col)
  }
  if (row < N - 1 && farm[row + 1][col] === 1 && !isVisit[row + 1][col]) {
    dfs(row + 1, col)
  }
  if (col > 0 && farm[row][col - 1] === 1 && !isVisit[row][col - 1]) {
    dfs(row, col - 1)
  }
  if (col < M - 1 && farm[row][col + 1] === 1 && !isVisit[row][col + 1]) {
    dfs(row, col + 1)
  }
}

File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/8. 1012.py
from collections import deque
import sys

sys.setrecursionlimit(10000)

input = sys.stdin.readline
output = sys.stdout.write

T = int(input())

# bfs 함수 내에서 global 로 쓰기 위해 미리 선언
N = 1
M = 1
K = 1

farm = []
isVisit = []

visitQueue = deque()

def bfs(row, col):
  global N, M
  
  if row > 0 and farm[row - 1][col] == 1 and not isVisit[row - 1][col]:
    isVisit[row - 1][col] = True
    visitQueue.append([row - 1, col])
  if row < N - 1 and farm[row + 1][col] == 1 and not isVisit[row + 1][col]:
    isVisit[row + 1][col] = True
    visitQueue.append([row + 1, col])
  if col > 0 and farm[row][col - 1] == 1 and not isVisit[row][col - 1]:
    isVisit[row][col - 1] = True
    visitQueue.append([row, col - 1])
  if col < M - 1 and farm[row][col + 1] == 1 and not isVisit[row][col + 1]:
    isVisit[row][col + 1] = True
    visitQueue.append([row, col + 1])

  if len(visitQueue) > 0:
    nextRow, nextCol = visitQueue.popleft()
    bfs(nextRow, nextCol)

while T > 0:
  m, n, k = map(int, input().split())

  N = n
  M = m
  K = k

  farm = list(list(0 for i in range(M)) for j in range(N))
  isVisit = list(list(False for i in range(M)) for j in range(N))

  for i in range(K):
    x, y = map(int, input().split())
    farm[y][x] = 1

  result = 0

  for row in range(N):
    for col in range(M):
      if farm[row][col] == 1 and not isVisit[row][col]:
        isVisit[row][col] = True
        bfs(row, col)
        result += 1

  output(f"{result}\n")

  T -= 1

